### ISSUE TITLE
Errors in variables?

### ISSUE THREAD

**Issue by m93a** [mljs/levenberg-marquardt#17]
I need to fit data points with errors in both variables. Does a "weighted" version of this algorithm exist? Would it be viable to implement it? I've spent hours searching for an implementation of something like that (in any understandable language) but had little success.

**Comment by maasencioh** [comment]
Hello @m93a what do you mean with weighted?

**Comment by m93a** [comment]
I mean weighted as in "Weighted Total Least Squares", which apart from values for `x` and `y` needs their standard deviations. When computing the regression, it gives more weight to data points with lower error (standard deviation) and vice versa. With TLS, the modification to the weighted form is as simple as multiplying the residuals by the reciprocal of error. But I don't understand the technical aspects of the Levenberg-Marquardt algorithm so I don't know whether something like this is also possible.

**Comment by jobo322** [comment]
Hello @m93a i'm interested in your approach, I am trying to understand your explanation, when you say merit function, this is inbebed in the [step function](https://github.com/mljs/levenberg-marquardt/blob/master/src/step.js). There is a function called gradient function, this creates a gradient matrix used to compute the new candidates, the matrixFunction return the difference matrix between toFit data ('real' data) and the Func(xi, B) result where B is the vector of current parameters. between both function could you find the place to implement your idea.

**Comment by andcastillo** [comment]
Hello there,

Indeed, the LM algorithm should minimize sum of **weighted** squared residuals. I remember in the original implementation there were a parameter called _weight = weighting vector for least squares fit ( weight >= 0 ) ..._ (https://github.com/mljs/curve-fitting)[https://github.com/mljs/curve-fitting]
This does not actually solve all your problem, because, the LM algorithm consider the independent variable free of error, so σ( f(xᵢ, β) ) = 0. Then, as far as I understand, you can deal with the error σ(yᵢ) by means of the weight vector, but somehow also with the error in  xᵢ becuse yᵢ = g(xi): it is an implicit function of xᵢ.
Now, this version of the LM lacks the implementation of the weight, but that is almost trivial having the old code in the curve fitting project, so perhaps you could do it.
As I said, it could be good approximation to the solution. If not, you will need to change LM to work with function h(i) = <y(i), x(i)>, which is not supported by any implementation I currently know.

**Comment by m93a** [comment]
Hey again,
Thank you for your comments and tips!

I've looked at the old code of `ml-curve-fitting`, as @andcastillo suggested, and found [the merit function](https://github.com/mljs/curve-fitting/blob/master/src/LM.js#L198) very quickly. However when I tried to understand the context, I found the code quite hard to follow, as it lacked types, segmentation to sub-functions and descriptive names.

Code in *this repo*, on the other hand, is rather delightful and bears little resemblance to `ml-curve-fitting` (probably for the better), however I failed to find a part that would resemble the merit function, so I took @jobo322's advice and looked into the `gradientFunction`.

If I understand it correctly, the gradient function returns the gradient of residuals:
```
                  ⎛ ∂/∂β₁ ⎞   ⎛ y₁ − f(x₁, β) ⎞
                  ⎜ ∂/∂β₂ ⎟   ⎜ y₂ − f(x₂, β) ⎟
      G = ∇ ⊗ R = ⎜  ...  ⎟ ⊗ ⎜      ...      ⎟
                  ⎝ ∂/∂βₘ ⎠   ⎝ yₙ − f(xₙ, β) ⎠
```
where `R` is the vector of residuals, `∂/∂βᵢ` is the partial derivative with respect to parameter `βᵢ` and `a ⊗ b` is the outer product of two vectors. If the gradient `G` minimizes the norm of `R`, simply weighting every entry of `R` by the reciprocal of the error should produce the correct result, right? In code that would mean adding something like `ans[param][point] /= dataErrors[point]` to [here](https://github.com/mljs/levenberg-marquardt/blob/master/src/step.js#L32).

Now there's the problem with determining the error… In the book *Numerical Recipes in FORTRAN 77* (page 660 in book, page 690 in [pdf](https://websites.pmc.ucsc.edu/~fnimmo/eart290c_17/NumericalRecipesinF77.pdf#page=690&zoom=auto,-187,449)) they say that the statistically correct way to fit a straight line to data with errors in variables is to minimize this value:
```
              ( yᵢ − a − bxᵢ )²
      χ² = Σ ———————————————————
              σ(yᵢ)² + b²σ(xᵢ)²
```
This exactly corresponds to the equation I've given in my previous comment if you substitute `f(x)` for `a + bx`. I don't have the knowledge to verify that the book is actually correct, but I strongly believe that it is. This would mean that my equation for the weights is presumably correct and the only thing left is to determine the value of `σ( f(xᵢ, β) )` from `xᵢ` and `σ(xᵢ)`…

Since the equation `σ(f(xᵢ)) ≈ (∂f/∂x)(xᵢ) * σ(xᵢ)` applies only for small `σ(xᵢ)` and I don't want to risk hitting a point where `f(xᵢ)` is locally very flat but gets very steep as you move to `f(xᵢ + σ)`, I'll try to solve it in a different way:

We assume that the actual value (let's call it `ξᵢ`) which was measured as `xᵢ` lays somewhere around it with the probability given by the normal distribution with mean `xᵢ` and standard deviation `σᵢ`. This means that `ξᵢ` lays in the range between `xᵢ − 3σᵢ` and `xᵢ + 3σᵢ` with the probability of 99.7 %. I propose that we divide this range to `N` regions, so that each region has the same probability of containing `ξᵢ`. Now, if we evaluate `f` at the edge of each region, we should be able to use [the basic formula for standard deviation](https://www.mathsisfun.com/data/standard-deviation-formulas.html) to approximate it in the neighbourhood of `xᵢ` (and hopefully `ξᵢ`).

Wow, that was a lot of maths… If you either don't understand something, think I got something wrong or want to approve my conclusions, feel encouraged to comment!

**Comment by SimonEast** [comment]
Apologies for commenting on an 8 year old thread, but in case others come here looking for answers (as I did), I thought I might share my solution. This approach was suggested by Grok AI and appears to work well from my initial testing.

I'm not sure whether support for weights was fully integrated into the library (the pull request above seems to not yet be merged). But I found weighting relatively easy to implement in my main script - no modifications to the library itself were required.

See this interactive, working example:
https://stackblitz.com/edit/stackblitz-starters-dpweyzgp?file=index.js&view=editor

It uses equal weighting by default `[1, 1, 1]` but by adjusting those numbers, you can change how much each value gets adjusted by the algorithm. A value of `[1, 1, 2]` means the third parameter is trusted twice as much as the other two, and should be adjusted around half as much. Or a value of `[1, 1, 100000]` basically holds that third parameter steady and only adjusts the other two to fit.

I hope that makes sense and helps someone else when working on adjustment problems.
